src/NServiceBus.Core/Recoverability/FirstLevelRetries/FirstLevelRetriesHandler.cs

```diff
@@ -0,0 +1,46 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Runtime.ExceptionServices;
+    using System.Threading.Tasks;
+    using Logging;
+    using Pipeline;
+
+    class FirstLevelRetriesHandler
+    {
+        public FirstLevelRetriesHandler(FailureInfoStorage storage, FirstLevelRetryPolicy retryPolicy)
+        {
+            this.storage = storage;
+            this.retryPolicy = retryPolicy;
+        }
+
+        public async Task<bool> HandleMessageFailure(ITransportReceiveContext context, Exception ex)
+        {
+            var messageId = context.Message.MessageId;
+            var pipelineUniqueMessageId = messageId;
+
+            var firstLevelRetries = storage.GetFailureInfoForMessage(pipelineUniqueMessageId).FLRetries;
```

Comment:
I think that if we decide to move the query part to `RecoverabilityBehavior` we should move the whole failure persitence logic there including updating/clearing logic as well. That will make the `RecoverabiltyBehavior` a lot more complicated which I would prefer to avoid.